packages/go_router/lib/src/state.dart

```diff
@@ -81,4 +137,130 @@ class GoRouterState {
     return _configuration.namedLocation(name,
         params: params, queryParams: queryParams);
   }
+
+  @override
+  bool operator ==(Object other) {
+    return other is GoRouterState &&
+        other.location == location &&
+        other.subloc == subloc &&
+        other.name == name &&
+        other.path == path &&
+        other.fullpath == fullpath &&
+        other.params == params &&
+        other.queryParams == queryParams &&
+        other.queryParametersAll == queryParametersAll &&
+        other.extra == extra &&
+        other.error == error &&
+        other.pageKey == pageKey;
+  }
+
+  @override
+  int get hashCode => Object.hash(location, subloc, name, path, fullpath,
+      params, queryParams, queryParametersAll, extra, error, pageKey);
+}
+
+/// An inherited widget to host a [GoRouterStateRegistry] for the subtree.
+///
+/// Should not be used directly, consider using [GoRouterState.of] to access
+/// [GoRouterState] from the context.
+class GoRouterStateRegistryScope
+    extends InheritedNotifier<GoRouterStateRegistry> {
+  /// Creates a GoRouterStateRegistryScope.
+  const GoRouterStateRegistryScope({
+    super.key,
+    required GoRouterStateRegistry registry,
+    required super.child,
+  }) : super(notifier: registry);
+}
+
+/// A registry to record [GoRouterState] to [Page] relation.
+///
+/// Should not be used directly, consider using [GoRouterState.of] to access
+/// [GoRouterState] from the context.
+class GoRouterStateRegistry extends ChangeNotifier {
+  /// creates a [GoRouterStateRegistry].
+  GoRouterStateRegistry();
+
+  /// A [Map] that maps a [Page] to a [GoRouterState].
+  @visibleForTesting
+  final Map<Page<Object?>, GoRouterState> registry =
+      <Page<Object?>, GoRouterState>{};
+
+  final Set<Page<Object?>> _pagesWithAssociation = <Page<Object?>>{};
+  final Map<Route<Object?>, Page<Object?>> _routePageAssociation =
```

Comment:
Is this needed? It looks like `_pagesWithAssociation` is equivalent to `_routePageAssociation.values`.

Reply:
yes, but look up time is O(1) vs O(n)

Reply:
Should we consider memory usage or is it worth the tradeoff here?

Reply:
done. i removed `_pagesWithAssociation` and created in flight in updateRegistry.